Hybrid 1D/2D measurements with analysis_v2

stephlazar opened the pull request:
Changes proposed in this pull request:
- Upgrades BaseDataAnalysis.add_measured_data to handle the case of hybrid measurements:
   conceptually a 2D measurement that was compressed along the 1st sweep dimension and
   the measurement was run in 1D mode.
- Fixes label and unit extraction in MultiQubit_TimeDomain_Analysis plotting function to
   work with SweepPoints.

This feature will be used automatically when the SweepPoints object for the measurement has 2 dimensions, but there is only one column of sweep points in the HDF5 file.

Analysis should be run with the flag TwoD == True (pass in options_dict).

!!! This feature only works if the SweepPoints are passed.

@antsr FYI

chellings committed:
sweep_points: added possibility to initialize from a list of sweep dicts

Commit:
Upgrades BaseDataAnalysis.add_measured_data to handle the case of hybrid measurements:
   conceptually a 2D measurement that was compressed along the 1st sweep dimension and
   the measurement was run in 1D mode.
Fixes label and unit extraction in MultiQubit_TimeDomain_Analysis plotting function to
   work with SweepPoints.

nathlacroix commented on pycqed/analysis_v2/base_analysis.py:
```diff
@@ -363,6 +369,14 @@ def add_measured_data(raw_data_dict, compression_factor=1):
+            # (so only 1 column of sweep points in hdf5 file)
+            # CURRENTLY ONLY WORKS WITH SweepPoints OBJECTS
+            if cp is None:
+                cp = CalibrationPoints('qb', states=[])
```
qb_names should always be a list if the wrapper CalibrationPoints.single_qubit(...) is not used
otherwise it might result in odd behavior

stephlazar replied:
Ok got it!

nathlacroix commented on pycqed/analysis_v2/base_analysis.py:
```diff
@@ -427,7 +464,10 @@ def extract_data(self):
-                self.get_param_value('compression_factor', 1))
+                self.get_param_value('compression_factor', 1),
+                self.get_param_value('sweep_points'),
+                CalibrationPoints.from_string(
```
will fail if there is no cal point object (from string doesn't work with None). Should the default value be CalibrationPoints([], []) ?

stephlazar replied:
Thanks! I didn't check that.

nathlacroix commented on pycqed/analysis_v2/base_analysis.py:
```diff
@@ -372,7 +386,14 @@ def add_measured_data(raw_data_dict, compression_factor=1):
+                # deal with hybrid measurements
+                if sweep_points.shape[0] == 1 and len(sp) > 1:
+                    hybrid_measurement = True
+                    hsp = np.arange(len(next(iter(sp[0].values()))[0]))
```
quite compact but a bit hard to read, I must say :D 5 nested function calls + 2 indexing... maybe doing it in 2 steps would make it more readable? or a comment explaining what happens (just imagine somehow the inner structure of SweepPoints changes and someone is asked to check if the code would work here with the new structure, would be pretty hard to say just by looking at this one liner :p)

stephlazar replied:
This is done in many places throughout timedomain_analysis :p And basically every time I need to access some parameters from SweepPoints. I don't want to have 2 lines of code every time I need some parameters from SweepPoints.
It's true that here I take np.arange and len in addition to the next(iter()) so I will split it in 2 lines.

nathlacroix replied:
if you need it each time you have to access sweep points than the even nicer solution is to create a function within the sweeppoint object. This is best because also it hides inner structure of sweeppoints and in case the structure changes, we can just change that "getter" function. A bit like the "cal_points.get_states()"

stephlazar replied:
yes, that is a good solution. I will need to go through everywhere I did this but it's better to do it now than later after we will have used it in many more places :P

stephlazar replied:
I have done this now in an atomic commit with SHA-1: 92fcca176b94028c2470dbfd1ecb50bdfeb5bd10. We should cherry pick this commit into our paralle_2qb_calib_class branch so we use them there.

nathlacroix commented on pycqed/analysis_v2/base_analysis.py:
```diff
@@ -372,7 +386,14 @@ def add_measured_data(raw_data_dict, compression_factor=1):
+                if sweep_points.shape[0] == 1 and len(sp) > 1:
+                    hybrid_measurement = True
+                    hsp = np.arange(len(next(iter(sp[0].values()))[0]))
+                    ssp = np.arange(len(next(iter(sp[1].values()))[0]))
```
sp, ssp , sweep_points. I think those variable names are quite confusing. Maybe time to rename sweep_points to "mc_points" here also?

stephlazar replied:
Yes! But i will keep sp or? Just because it has a symmetry to cp which we use throughout timedomain_analysis.py

nathlacroix replied:
yes, I agree, we could keep sp, would rename "sweep_points" to mc_points

nathlacroix commented on pycqed/analysis_v2/base_analysis.py:
```diff
@@ -383,7 +404,23 @@ def add_measured_data(raw_data_dict, compression_factor=1):
+                            # add back ssl number of copies of the cal points
+                            # at the end of each soft sweep slice
+                            cal_pts = data[i][-len(cp.states):]
+                            cal_pts_arr = np.reshape(np.repeat(cal_pts, ssl),
```
Are you sure what you're doing here would work with e.g. active reset and preselection? those readout would not be included in SweepPoints, from which you retrieve the number of readouts for the reshaping, right? Also, as you already corrected somewhere else in the analysis, with preparation params the data of cal_points is not only len(cp.states).

stephlazar replied:
Good catch, you are right, it wouldn't....

stephlazar replied:
I fixed this now. I don't really have good data to test it on. I have some files from Ants' first tries where the data is just noise basically. The analysis works for these. After that he changed the SweepPoints such that even though he conceptually has a 2D measurement, he only saves one dict in SweepPoints. But my implementation relies on the SweepPoints being 2D.

nathlacroix reviewed (changes requested):
Thanks for the added functionality!
I have the impression though that it is not compatible with preparation parameters. 
It now also makes the base analysis aware (and hence dependent) of cal_points and sweep points, not sure whether this is really something we want or not...  what do you think?

Commit:
Merge branch 'qudev_master' into enh/analysisv2_hybrid_measurements

Commit:
Adds convenient getter methods to SweepPoints.

Adds methods get_sweep_dimension, get_sweep_params_description, get_sweep_params_property to
   SweepPoints.
Upgrades MultiQubit_TimeDomain_Analysis to use these new SweepPoints methods.

Commit:
Improves BaseDataAnalysis.add_measured_data() to take into account prep_params for hybrid measurements.

stephlazar commented:
Regarding the BaseDataAnalysis relying on calibration/sweep points + prep params. I don't think this is a problem for 2 reasons:

1. The class already extracts all of these and are in raw data dict. So in a sense it did have them the whole time. 
2. add_measured_data which now explicitly uses calibration/sweep points + prep params does not really depend on them. I give nice default values in case any of these are not passed to the analysis in some way.

nathlacroix commented on pycqed/analysis_v2/base_analysis.py:
```diff
@@ -362,19 +370,48 @@ def add_measured_data(raw_data_dict, compression_factor=1):
+                sp = SweepPoints(from_dict_list=sweep_points)
+                print(mc_points.shape[0], len(sp))
+                if mc_points.shape[0] == 1 and len(sp) > 1:
+                    print('here again')
```
debug print statement ;)

stephlazar replied:
Oops :p thanks!

nathlacroix reviewed (approved):
Thanks for the updates! Looks good to me now. 
Just (at least one) print debug statement that I found that probably should be removed before merging.

Commit:
Merge branch 'qudev_master' into enh/analysisv2_hybrid_measurements

Commit:
Removes debugging print statements in BaseDataAnalysis.
